### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by tal-sitton**

### CODE REVIEW

**Review by Yoel-Perman-Brilliant — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# SpikesLib - Examples
+Examples of how to use the [SpikesLib](https://github.com/Spikes-2212-Programming-Guild/SpikesLib2) library.
+
+## Includes:
+- Several subsystems and a drivetrain, each with its own command examples.
+- Each subsystem has a RobotMap class, which holds the ports for the robot's components.
+  - Each interface in RobotMap represents a different port.
```
Comment (OfriRosenbaum) [inline README.md]: a different type of port

File: README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# SpikesLib - Examples
+Examples of how to use the [SpikesLib](https://github.com/Spikes-2212-Programming-Guild/SpikesLib2) library.
+
+## Includes:
+- Several subsystems and a drivetrain, each with its own command examples.
```
Comment (OfriRosenbaum) [inline README.md]: and drivetrains

Comment (OfriRosenbaum) [inline README.md]: change `command` to `implementation`

File: README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+- Each subsystem has a RobotMap class, which holds the ports for the robot's components.
+  - Each interface in RobotMap represents a different port.
+
+## How To Use
+If you want to use the example code for yourself, simply copy & paste it, and make sure to change the ports in RobotMap,
+as well as the MotorController type when necessary.
```
Comment (OfriRosenbaum) [inline README.md:11]: as well as the type of the MotorController when necessary

File: src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    private static final double DISTANCE_PER_PULSE = 15.24 * Math.PI / 360.0 / 100;
+
+    private final Encoder leftEncoder, rightEncoder;
+    private final ADXRS450_Gyro gyro;
+
+    private final Supplier<Double> DRIVE_SPEED = rootNamespace.addConstantDouble("drive speed", 0.5);
+    private final Supplier<Double> METERS_TO_DRIVE = rootNamespace.addConstantDouble("meters to drive", 2);
+
+    /**
+     * A {@link Namespace} is an object which holds values on a {@link NetworkTable}. <br>
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java]: No need to explain again what a namespace is, already explained above

File: src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    private final Supplier<Double> kV = ffNamespace.addConstantDouble("kV", 0);
+    private final Supplier<Double> kA = ffNamespace.addConstantDouble("kA", 0);
+
+    private final FeedForwardSettings ffSettings = new FeedForwardSettings(kS, kV, kA);
+
+    /**
+     * The Drivetrain class is a singleton, which means it has only one instance. Since the constructor is private, a new
+     * instance cannot be instantiated, and you can only access the existing instance via the {@link #getInstance}
+     * function.
+     * Since the robot itself has only one drivetrain, we want a single instance of the class.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java]: we want **only** a single...

File: src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+     */
+    private static Drivetrain instance;
+
+    public static Drivetrain getInstance() {
+        if (instance == null)
+            instance = new Drivetrain();
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java]: remove blank line

File: src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    /**
+     * @return the robot's current angle in the range (-infinity, infinity)
+     */
+    public double getAngle() {
+        return gyro.getAngle();
+    }
+
+    /**
+     * @return the robot's current angle in the range (-180, 180]
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java:154]: [-180,180]

File: src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Robot.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    public void disabledInit() {
+    }
+
+    @Override
+    public void disabledPeriodic() {
+    }
+
+    @Override
+    public void autonomousInit() {
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Robot.java]: remove blank line

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+
+import java.util.function.Supplier;
+
+/**
+ * A climber subsystem which consists of an arm that is controlled by a VictorSPX. The arm
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: `{@link WPI_VictorSPX}`

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ * can move up and grab a bar, then down to pull the robot up; thus making the robot climb. <br>
+ * A game that this example subsystem can be used in is Rapid React 2022.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/climber/Climber.java:14]: put this in a paragraph

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * The speed in which the arms move down in order to take the robot up after they caught the bar.
+     */
+    public static final double DOWN_SPEED = -0.4;
+
+    /**
+     * The Climber class is a singleton, which means it has only one instance. Since the constructor is private, a new
+     * instance cannot be instantiated, and you can only access the existing instance via the {@link #getInstance()}
+     * function.
+     * Since the robot itself only has one climber, we want only one instance of the class.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: we want only a single instance...

File: src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java (hunk 5 above)
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/drivetrains/advancedtankdrivetrain/Drivetrain.java]: we want only a single instance...

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+public class Gripper extends MotoredGenericSubsystem {
+
+    public static final double INTAKE_SPEED = 0.5;
+    public static final double DISPENSE_SPEED = -0.5;
+
+    /**
+     * The Gripper class is a singleton, which means it has only one instance. Since the constructor is private, a new
+     * instance cannot be instantiated, and you can only access the existing instance via the {@link #getInstance()}
+     * function. <br>
+     * Since the robot itself only has one gripper, we want only one instance of the class.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: we want only a single instance...

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj.motorcontrol.MotorController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+/**
+ * A gripper subsystem which consists of two motors, that both rotate either inwards or
+ * outwards. If they both rotate inwards, the gripper intakes a cargo. If they both rotate outwards, the gripper
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: change cargo to ball

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+ * dispenses the cargo.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: image

File: src/main/java/frc/robot/subsystems/gripper/Robot.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  @Override
+  public void robotPeriodic() {
+    gripper.periodic();
+    CommandScheduler.getInstance().run();
+  }
+
+  @Override
+  public void disabledInit() {}
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/gripper/Robot.java]: put the second curly brace on the line after this (do this for the other methods as well)

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        return instance;
+    }
+
+    private Gripper(MotorController motorController1, MotorController motorController2) {
+        super(DISPENSE_SPEED, INTAKE_SPEED, "gripper", motorController1, motorController2);
+        motorController2.setInverted(true);
+    }
+
+    /**
+     * Adds any commands or data from this subsystem to the {@link SmartDashboard}.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: change `SmartDashboard` to `Shuffleboard`

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.function.Supplier;
+
+/**
+ * A shooter subsystem that is controlled with two {@link WPI_TalonSRX}.<br>
+ * In addition, it has settings for PID and matching commands, which you can use and control through the {@link Shuffleboard}.
+ */
+public class Shooter extends MotoredGenericSubsystem {
+
+    /**
+     * Amount of rotations the wheel does for each encoder's pulse.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: change `Amount` to `Number`

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Places the PID constants on the {@link Shuffleboard}.
+     */
+    private final Supplier<Double> kP = pidNamespace.addConstantDouble("kP", 1);
+    private final Supplier<Double> kI = pidNamespace.addConstantDouble("kI", 0);
+    private final Supplier<Double> kD = pidNamespace.addConstantDouble("kD", 0);
+    private final Supplier<Double> TOLERANCE = pidNamespace.addConstantDouble("tolerance", 0);
+    private final Supplier<Double> WAIT_TIME = pidNamespace.addConstantDouble("wait time", 1);
+    private final Supplier<Double> targetVelocity = pidNamespace.addConstantDouble("target speed", 60);
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:53]: add javadoc (`Places the FeedForward constants on the {@link Shuffleboard}`)

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    private final Supplier<Double> kV = ffNamespace.addConstantDouble("kV", 0);
+
+    public final PIDSettings pidSettings = new PIDSettings(kP, kI, kD, TOLERANCE, WAIT_TIME);
+    public final FeedForwardSettings ffSettings = new FeedForwardSettings(kS, kV, () -> 0.0);
+
+    /**
+     * The Shooter class is a singleton, which means it has only one instance. Since the constructor is private, a new
+     * instance cannot be instantiated, and you can only access the existing instance via the {@code getInstance()}
+     * function.
+     * Since the robot itself has only one shooter, we want a single instance of the class.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: we want **only** a single instance of this class

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+    private final Supplier<Double> kS = ffNamespace.addConstantDouble("kS", 0);
+    private final Supplier<Double> kV = ffNamespace.addConstantDouble("kV", 0);
+
+    public final PIDSettings pidSettings = new PIDSettings(kP, kI, kD, TOLERANCE, WAIT_TIME);
+    public final FeedForwardSettings ffSettings = new FeedForwardSettings(kS, kV, () -> 0.0);
+
+    /**
+     * The Shooter class is a singleton, which means it has only one instance. Since the constructor is private, a new
+     * instance cannot be instantiated, and you can only access the existing instance via the {@code getInstance()}
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: change to `@code` to `@link` and add a reference to this method

File: src/main/java/frc/robot/subsystems/transfer/Robot.java [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    public void disabledInit() {
+    }
+
+    @Override
+    public void disabledPeriodic() {
+    }
+
+    @Override
+    public void autonomousInit() {
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/transfer/Robot.java]: remove blank line

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

**Review by OfriRosenbaum — APPROVED** [review]

**Comment by Yoel-Perman-Brilliant** [comment]
This was a triumph
I'm making a note here; "Huge success"

**Pull request merged**